test/ruby/test_hash.rb

```diff
@@ -122,6 +122,24 @@ def test_dup_will_rehash
     assert_equal set2, set2.dup
   end
 
+  def test_select_reject_will_not_rehash
+    x = []
+    set1 = @cls[x, true]
+    x << 42 # set1 need to be rehashed
+    set2 = set1.select { true }
+    set3 = set1.reject { false }
+    set1.rehash
+
+    assert_not_equal set1, set2
+    assert_not_equal set1, set3
```

Comment:
I'm afraid if we can guarantee these NEVER match.
While the same values have the same hash value, the reverse may not true however.

Reply:
I'm not sure a collision is even possible this way, but assuming it is, I think we can live with a spurious failure 1 every 2 ** 63 runs, no?

Reply:
@marcandre If you look at this another way, what is the value in testing that the sets are not equal before rehashing?  We can leave that behavior unspecified, and just keep the tests that they are equal after rehashing.

Put another way, let's say we changed the set implementation such that hashing was no longer needed, and rehash was made a no-op.  Would we want that to cause a failure in this test?

Reply:
@jeremyevans 

1) Note that the variable is named `set` but the class is `Hash`, not `Set`.

2) The value is to test that the operation "will not rehash". An equivalent test would be to test the change in `hash` value. Put another way, this test fails before this commit and passes after, i.e. red/green testing. 

3) The test that they are equal after rehashing is not part of the test, it's just sanity checking that the test itself is sound (and that the values are unequal strictly because one had it's hash out of sync).

4) "if we changed the set implementation" => Both `Hash` and `Set` implementation is specified; both `Hash` and `Set` rely on `Object#hash` and if elements are mutated, one needs to call `rehash`/`reset`. I don't see why or how that could ever change, but assuming that I am mistaken, than yes, I am fine if this test needed to be changed.

Reply:
@marcandre If the purpose is to test whether the hash is recalculated by select/reject, I think it's best to actually test that directly instead of testing the equality or inequality of the resulting collection object and using rehash. Something like:

```ruby
x = Object.new
hash_called = []
x.singleton_class.attr_writer :hash
x.define_singleton_method(:hash){hash_called << @hash; @hash}
x.hash = 1
set1 = @cls[x, true]
assert_equal([1], hash_called)
x.hash = 2
set1.select { true }
set1.reject { false }
assert_equal([1], hash_called)
```

Reply:
@jeremyevans that is a superior way, thank you! I should have thought of that 😅